### ISSUE TITLE
Differenciation of product of function

### ISSUE THREAD

**Issue by jeammimi** [google/jax#2667]
I am differentiating a function which is the product of a "lot" of function (up to 200), using np.prod .
So I was wondering how jax deal with it:
For example with three function it must compute
![$\frac{df_1}{dx}f2(x)f3(x) + f1(x) \frac{df_2}{dx}f3(x)+f1(x)f2(x)\frac{df_3}{dx}$](https://render.githubusercontent.com/render/math?math=%24%5Cfrac%7Bdf_1%7D%7Bdx%7Df2(x)f3(x)%20%2B%20f1(x)%20%5Cfrac%7Bdf_2%7D%7Bdx%7Df3(x)%2Bf1(x)f2(x)%5Cfrac%7Bdf_3%7D%7Bdx%7D%24)

So in this case f1 f2 and f3 are evaluated twice.
But with n functions they are evaluated n-1 times.

So basically it scale as n**2 evaluation, but could be managed as n evalutaion.

As my code is very long to evaluate for the first time , and also to compile it, I am wordering if it comes from here.

So the best way would be to have a vector f = (f1,f2,f3) which could be evaluated at x,
and also evaluate df/dx at x, and then combine it, but I am not sure how to do it in a efficient numpy vector way.
Best!

**Comment by hawkinsp** [comment]
Can you please provide a runnable reproduction? It's much easier to give concrete performance advice with a small, self-contained example that we can all run and discuss.

**Comment by gnecula** [comment]
Have you experienced this behaviour, or are you just thinking hypothetically. JAX includes sophisticated logic to try to avoid such unnecessary recomputation. 
Here is an example: 
```
    # 4 functions : x ** 1, x ** 4, ...
    funcs = [lambda x: x ** (1 + i * 3) for i in range(4)]
    def prod_func(x):
      return functools.reduce(lambda acc, f: acc * f(x), funcs, 1.)

    print(api.make_jaxpr(api.grad(prod_func))(3.))
```

The last line shows you the actual computation of the gradient. There are 4 invocations of `pow` for 4 functions (one to compute the function, and one to compute the derivative). 

You can replace `functools.reduce` with a loop, or writing the multiplication of functions by hand. The result will be the same.

**Comment by jeammimi** [comment]
I have not experienced this behaviour, I was just wondering  how it works.
When trying to submit a minimal example, I realised
that in the function that return the product I had a for loop that i could change by a vmap,
And then, the time to compute the gradient went from impossible to about 1 s.

I changed
`np.prod([fi(chi) for chi in chs])`

by
```
fim = vmap(fi,(0),0)
np.prod(fim(chs))
```

So I will close it
Best!
